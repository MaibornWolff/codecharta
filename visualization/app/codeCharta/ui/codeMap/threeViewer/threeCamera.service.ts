import { Injectable } from "@angular/core"
import { PerspectiveCamera, Vector3 } from "three"
import { EventEmitter } from "../../../util/EventEmitter"

type CameraChangeEvents = {
    onCameraZoomChanged: (data: { camera: PerspectiveCamera }) => void
}

@Injectable({ providedIn: "root" })
export class ThreeCameraService {
    static VIEW_ANGLE = 45
    static NEAR = 50
    static FAR = 200_000 //TODO optimize renderer for far objects
    static readonly MAX_ZOOM_Factor = 50
    static readonly MIN_ZOOM_Factor = 0.1
    static readonly ZOOM_STEP = 0.25
    private eventEmitter = new EventEmitter<CameraChangeEvents>()
    camera: PerspectiveCamera
    zoomFactor: number

    init(containerWidth: number, containerHeight: number) {
        const aspect = containerWidth / containerHeight
        this.camera = new PerspectiveCamera(ThreeCameraService.VIEW_ANGLE, aspect, ThreeCameraService.NEAR, ThreeCameraService.FAR)
        this.setPosition(new Vector3(0, 300, 1000))
        this.zoomFactor = this.camera.zoom
    }

    setPosition(position: Vector3) {
        this.camera.position.set(position.x, position.y, position.z)
        this.camera.lookAt(0, 0, 0)
    }

    setZoomFactor(zoomFactor: number) {
        this.camera.zoom = zoomFactor
        this.zoomFactor = zoomFactor
        this.eventEmitter.emit("onCameraZoomChanged", { camera: this.camera })
        this.camera.updateProjectionMatrix()
    }

    zoomIn() {
        this.setZoomFactor(Math.min(this.zoomFactor + ThreeCameraService.ZOOM_STEP, ThreeCameraService.MAX_ZOOM_Factor))
    }

    zoomOut() {
        this.setZoomFactor(Math.max(this.zoomFactor - ThreeCameraService.ZOOM_STEP, ThreeCameraService.MIN_ZOOM_Factor))
    }

    subscribe<Key extends keyof CameraChangeEvents>(key: Key, callback: CameraChangeEvents[Key]) {
        this.eventEmitter.on(key, data => {
            callback(data)
        })
    }
}
