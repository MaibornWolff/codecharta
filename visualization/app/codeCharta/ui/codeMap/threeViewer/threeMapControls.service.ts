import { ThreeCameraService } from "./threeCamera.service"
import { Injectable } from "@angular/core"
import { Box3, Mesh, MeshNormalMaterial, PerspectiveCamera, Vector3, Sphere, BoxGeometry, MOUSE } from "three"
import { ThreeSceneService } from "./threeSceneService"
import { MapControls } from "three/examples/jsm/controls/MapControls"
import { ThreeRendererService } from "./threeRenderer.service"
import { EventEmitter } from "../../../util/EventEmitter"
import { BehaviorSubject } from "rxjs"

type CameraChangeEvents = {
    onCameraChanged: (data: { camera: PerspectiveCamera }) => void
}

@Injectable({ providedIn: "root" })
export class ThreeMapControlsService {
    static CAMERA_CHANGED_EVENT_NAME = "camera-changed"
    MAX_ZOOM = 200
    MIN_ZOOM = 10

    controls: MapControls
    private eventEmitter = new EventEmitter<CameraChangeEvents>()
    zoomPercentage$ = new BehaviorSubject<number>(100)
    positionBeforeFocus: Vector3

    constructor(
        private threeCameraService: ThreeCameraService,
        private threeSceneService: ThreeSceneService,
        private threeRendererService: ThreeRendererService
    ) {}

    setControlTarget(cameraTarget: Vector3) {
        this.controls.target.set(cameraTarget.x, cameraTarget.y, cameraTarget.z)
    }

    rotateCameraInVectorDirection(x: number, y: number, z: number) {
        const zoom = this.getZoom()
        this.lookAtDirectionFromTarget(x, y, z)
        this.applyOldZoom(zoom)

        this.threeRendererService.render()
        this.onInput(this.threeCameraService.camera)
    }

    autoFitTo() {
        setTimeout(() => {
            const boundingSphere = this.getBoundingSphere()
            if (!boundingSphere || boundingSphere.radius === -1) {
                return
            }

            const length = this.cameraPerspectiveLengthCalculation(boundingSphere)
            const cameraReference = this.threeCameraService.camera

            cameraReference.position.set(length, length, boundingSphere.center.z)
            this.positionBeforeFocus = new Vector3(length, length, boundingSphere.center.z)

            this.controls.update()

            this.focusCameraViewToCenter(boundingSphere)
            this.threeRendererService.render()
            this.onInput(this.threeCameraService.camera)

            const scale = 1.3 // object size / display size

            this.controls.maxDistance = length * 4
            this.controls.minDistance = boundingSphere.radius / (10 * scale)

            this.setZoomPercentage(140)
        })
    }

    private cameraPerspectiveLengthCalculation(boundingSphere: Sphere) {
        const cameraReference = this.threeCameraService.camera

        //TODO: Scale Factor for object to camera ratio
        const scale = 1.3 // object size / display size
        const objectAngularSize = ((cameraReference.fov * Math.PI) / 180) * scale

        const distanceToCamera = boundingSphere.radius / Math.tan(objectAngularSize / 2)
        return Math.sqrt(Math.pow(distanceToCamera, 2) + Math.pow(distanceToCamera, 2))
    }

    private focusCameraViewToCenter(boundingSphere: Sphere) {
        const boundingSphereCenter: Vector3 = boundingSphere.center.clone()

        boundingSphereCenter.setY(0)

        this.controls.target.set(boundingSphereCenter.x, boundingSphereCenter.y, boundingSphereCenter.z)

        this.threeCameraService.camera.lookAt(boundingSphereCenter)

        this.threeCameraService.camera.updateProjectionMatrix()
    }

    getBoundingSphere() {
        return new Box3().setFromObject(this.threeSceneService.mapGeometry).getBoundingSphere(new Sphere())
    }

    private lookAtDirectionFromTarget(x: number, y: number, z: number) {
        this.threeCameraService.camera.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

        const alignmentCube = new Mesh(new BoxGeometry(20, 20, 20), new MeshNormalMaterial())

        this.threeSceneService.scene.add(alignmentCube)

        alignmentCube.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

        alignmentCube.translateX(x)
        alignmentCube.translateY(y)
        alignmentCube.translateZ(z)

        this.threeCameraService.camera.lookAt(alignmentCube.getWorldPosition(alignmentCube.position))
        this.threeSceneService.scene.remove(alignmentCube)
    }

    private getZoom() {
        return this.threeCameraService.camera.position.distanceTo(this.controls.target)
    }

    private applyOldZoom(oldZoom: number) {
        this.threeCameraService.camera.translateZ(oldZoom)
    }

    init(domElement: HTMLCanvasElement) {
        this.controls = new MapControls(this.threeCameraService.camera, domElement)
        this.controls.mouseButtons = {
            LEFT: MOUSE.ROTATE,
            MIDDLE: MOUSE.DOLLY,
            RIGHT: MOUSE.PAN
        }

        window.addEventListener("wheel", event => {
            this.controls.zoomToCursor = event.deltaY <= 0
        })

        this.controls.listenToKeyEvents(window)
        this.controls.addEventListener("change", () => {
            this.onInput(this.threeCameraService.camera)
            this.updateZoomPercentage()
            this.threeRendererService.render()
        })
        this.updateZoomPercentage()
    }

    onInput(camera: PerspectiveCamera) {
        this.setControlTarget(this.controls.target)
        this.eventEmitter.emit("onCameraChanged", { camera })
    }

    subscribe<Key extends keyof CameraChangeEvents>(key: Key, callback: CameraChangeEvents[Key]) {
        this.eventEmitter.on(key, data => {
            callback(data)
        })
    }

    getZoomPercentage(distance: number): number {
        const min = this.controls.minDistance
        const max = this.controls.maxDistance

        if (distance <= min) {
            return this.MAX_ZOOM
        }
        if (distance >= max) {
            return this.MIN_ZOOM
        }

        const range = max - min
        return this.MAX_ZOOM - ((distance - min) / range) * (this.MAX_ZOOM - this.MIN_ZOOM)
    }

    getDistanceFromZoomPercentage(percentage: number): number {
        const min = this.controls.minDistance
        const max = this.controls.maxDistance
        const range = max - min

        return min + ((this.MAX_ZOOM - percentage) / (this.MAX_ZOOM - this.MIN_ZOOM)) * range
    }

    updateZoomPercentage() {
        const distance = this.threeCameraService.camera.position.distanceTo(this.controls.target)
        const zoomFactor = this.getZoomPercentage(distance)
        this.zoomPercentage$.next(zoomFactor)
    }

    setZoomPercentage(zoom: number) {
        const newDistance = this.getDistanceFromZoomPercentage(zoom)
        const direction = new Vector3().subVectors(this.threeCameraService.camera.position, this.controls.target).normalize()
        this.threeCameraService.camera.position.copy(this.controls.target).add(direction.multiplyScalar(newDistance))
        this.controls.update()

        this.zoomPercentage$.next(zoom)
    }

    focusNode(nodePath: string) {
        this.positionBeforeFocus = this.threeCameraService.camera.position.clone()
        const mapMesh = this.threeSceneService.getMapMesh()
        const node = mapMesh.getBuildingByPath(nodePath)
        const boundingSphere = node.boundingBox.getBoundingSphere(new Sphere())

        this.ensureProperDistanceAndFocus(boundingSphere)
    }

    unfocusNode(callback?: () => void): void {
        if (!this.positionBeforeFocus) {
            return
        }

        const startPos = this.threeCameraService.camera.position.clone()
        const endPos = this.positionBeforeFocus.clone()
        const startTime = performance.now()
        const duration = 1000 // 1 second for the transition

        const animate = (currentTime: number) => {
            const elapsedTime = currentTime - startTime
            const t = Math.min(elapsedTime / duration, 1)

            const newPosition = new Vector3().lerpVectors(startPos, endPos, t)
            this.moveCameraToPosition(newPosition, this.controls.target, false)

            if (t < 1) {
                requestAnimationFrame(animate)
            } else if (callback) {
                callback()
            }
        }

        requestAnimationFrame(animate)
    }

    private ensureProperDistanceAndFocus(boundingSphere: Sphere) {
        const currentDistance = this.threeCameraService.camera.position.distanceTo(boundingSphere.center)
        const requiredDistance = this.cameraPerspectiveLengthCalculation(boundingSphere) * 1.8

        if (currentDistance < requiredDistance) {
            const zoomOutPosition = this.calculateCameraEndPosition(boundingSphere.center, boundingSphere.radius, false)
            this.animateCameraTransition({ center: zoomOutPosition, radius: requiredDistance } as Sphere, 500, () => {
                this.animateCameraTransition(boundingSphere, 1000)
            })
        } else {
            this.animateCameraTransition(boundingSphere, 1000)
        }
    }

    private animateCameraTransition(boundingSphere: Sphere, duration: number, callback?: () => void) {
        const { center, radius } = boundingSphere

        const startPos = this.threeCameraService.camera.position.clone()
        const endPos = this.calculateCameraEndPosition(center, radius, true)

        const startTime = performance.now()

        const animate = (currentTime: number) => {
            const elapsedTime = currentTime - startTime
            const t = Math.min(elapsedTime / duration, 1)

            const newPosition = new Vector3().lerpVectors(startPos, endPos, t)
            this.moveCameraToPosition(newPosition, center, false)

            if (t < 1) {
                requestAnimationFrame(animate)
            } else if (callback) {
                callback()
            }
        }

        requestAnimationFrame(animate)
    }

    private calculateCameraEndPosition(center: Vector3, radius: number, isFocusing: boolean): Vector3 {
        const currentCameraPosition = this.threeCameraService.camera.position.clone()
        const direction = this.controls.target.clone().sub(currentCameraPosition).normalize()

        const distance = this.cameraPerspectiveLengthCalculation({ center, radius } as Sphere) * (isFocusing ? 1.8 : 1)

        return center.clone().sub(direction.multiplyScalar(distance))
    }

    private moveCameraToPosition(position: Vector3, target: Vector3, render = true) {
        const camera = this.threeCameraService.camera

        camera.position.copy(position)
        camera.lookAt(target)

        this.controls.target.copy(target)
        this.controls.update()

        if (render) {
            this.threeRendererService.render()
        }

        this.onInput(camera)
    }
}
