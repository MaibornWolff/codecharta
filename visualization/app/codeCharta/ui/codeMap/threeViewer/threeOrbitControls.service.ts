import { ThreeCameraService } from "./threeCamera.service"
import { Injectable } from "@angular/core"
import { Box3, Mesh, MeshNormalMaterial, PerspectiveCamera, Vector3, Sphere, BoxGeometry } from "three"
import { ThreeSceneService } from "./threeSceneService"

import type { OrbitControls as OrbitControlsType } from "three/examples/jsm/controls/OrbitControls"
// TODO remove this old orbital control and use the jsm examples oneW
// eslint-disable-next-line no-duplicate-imports
import * as Three from "three"
import oc from "three-orbit-controls"
import { ThreeRendererService } from "./threeRenderer.service"
import { EventEmitter } from "../../../util/EventEmitter"

type CameraChangeEvents = {
    onCameraChanged: (data: { camera: PerspectiveCamera }) => void
}

@Injectable({ providedIn: "root" })
export class ThreeOrbitControlsService {
    static CAMERA_CHANGED_EVENT_NAME = "camera-changed"

    controls: OrbitControlsType
    private eventEmitter = new EventEmitter<CameraChangeEvents>()

    constructor(
        private threeCameraService: ThreeCameraService,
        private threeSceneService: ThreeSceneService,
        private threeRendererService: ThreeRendererService
    ) {}

    setControlTarget(cameraTarget: Vector3) {
        this.controls.target.set(cameraTarget.x, cameraTarget.y, cameraTarget.z)
    }

    rotateCameraInVectorDirection(x: number, y: number, z: number) {
        const zoom = this.getZoom()
        this.lookAtDirectionFromTarget(x, y, z)
        this.applyOldZoom(zoom)

        this.threeRendererService.render()
        this.onInput(this.threeCameraService.camera)
    }

    autoFitTo() {
        setTimeout(() => {
            const boundingSphere = this.getBoundingSphere()

            const length = this.cameraPerspectiveLengthCalculation(boundingSphere)
            const cameraReference = this.threeCameraService.camera

            cameraReference.position.set(length, length, boundingSphere.center.z)

            this.controls.update()

            this.focusCameraViewToCenter(boundingSphere)
            this.threeRendererService.render()
            this.onInput(this.threeCameraService.camera)
        })
    }

    private cameraPerspectiveLengthCalculation(boundingSphere: Sphere) {
        const cameraReference = this.threeCameraService.camera

        //TODO: Scale Factor for object to camera ratio
        const scale = 1.3 // object size / display size
        const objectAngularSize = ((cameraReference.fov * Math.PI) / 180) * scale

        const distanceToCamera = boundingSphere.radius / Math.tan(objectAngularSize / 2)
        return Math.sqrt(Math.pow(distanceToCamera, 2) + Math.pow(distanceToCamera, 2))
    }

    private focusCameraViewToCenter(boundingSphere: Sphere) {
        const boundingSphereCenter: Vector3 = boundingSphere.center.clone()

        boundingSphereCenter.setY(0)

        this.controls.target.set(boundingSphereCenter.x, boundingSphereCenter.y, boundingSphereCenter.z)

        this.threeCameraService.camera.lookAt(boundingSphereCenter)

        this.threeCameraService.camera.updateProjectionMatrix()
    }

    private getBoundingSphere() {
        return new Box3().setFromObject(this.threeSceneService.mapGeometry).getBoundingSphere(new Sphere())
    }

    private lookAtDirectionFromTarget(x: number, y: number, z: number) {
        this.threeCameraService.camera.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

        const alignmentCube = new Mesh(new BoxGeometry(20, 20, 20), new MeshNormalMaterial())

        this.threeSceneService.scene.add(alignmentCube)

        alignmentCube.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

        alignmentCube.translateX(x)
        alignmentCube.translateY(y)
        alignmentCube.translateZ(z)

        this.threeCameraService.camera.lookAt(alignmentCube.getWorldPosition(alignmentCube.position))
        this.threeSceneService.scene.remove(alignmentCube)
    }

    private getZoom() {
        return this.threeCameraService.camera.position.distanceTo(this.controls.target)
    }

    private applyOldZoom(oldZoom: number) {
        this.threeCameraService.camera.translateZ(oldZoom)
    }

    init(domElement: HTMLCanvasElement) {
        const OrbitControls = oc(Three)
        this.controls = new OrbitControls(this.threeCameraService.camera, domElement) as unknown as OrbitControlsType
        this.controls.addEventListener("change", () => {
            this.onInput(this.threeCameraService.camera)
        })
    }

    onInput(camera: PerspectiveCamera) {
        this.setControlTarget(this.controls.target)
        this.eventEmitter.emit("onCameraChanged", { camera })
    }

    subscribe<Key extends keyof CameraChangeEvents>(key: Key, callback: CameraChangeEvents[Key]) {
        this.eventEmitter.on(key, data => {
            callback(data)
        })
    }
}
