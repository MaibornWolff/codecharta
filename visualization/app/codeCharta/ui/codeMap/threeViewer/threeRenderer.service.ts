import { Injectable } from "@angular/core"
import { Camera, RGBAFormat, Scene, Vector2, WebGLInfo, WebGLRenderer, WebGLRenderTarget } from "three"
import { RenderPass } from "three/examples/jsm/postprocessing/RenderPass"
import WEBGL from "three/examples/jsm/capabilities/WebGL"
import { ShaderPass } from "three/examples/jsm/postprocessing/ShaderPass"
import { CustomComposer } from "../rendering/postprocessor/customComposer"
import { isWhiteBackgroundSelector } from "../../../state/store/appSettings/isWhiteBackground/isWhiteBackground.selector"
import { SharpnessMode, CcState } from "../../../codeCharta.model"
import { fxaaShaderStrings } from "../rendering/shaders/loaders/fxaaShaderStrings"
import { Store, State } from "@ngrx/store"

@Injectable({ providedIn: "root" })
export class ThreeRendererService {
    static BACKGROUND_COLOR = {
        white: 0xff_ff_ff,
        normal: 0xf4_f4_eb
    }

    static CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.normal

    static CLEAR_ALPHA = 1

    static RENDER_OPTIONS: WebGLContextAttributes = {
        antialias: true,
        preserveDrawingBuffer: true,
        alpha: true
    }

    static enableFXAA = false
    static setPixelRatio = false

    static instance: ThreeRendererService

    composer: CustomComposer
    renderer: WebGLRenderer
    scene: Scene
    camera: Camera

    constructor(
        private store: Store<CcState>,
        private state: State<CcState>
    ) {}

    init(containerWidth: number, containerHeight: number, scene: Scene, camera: Camera) {
        this.scene = scene
        this.camera = camera
        this.initGL(containerWidth, containerHeight)
        this.store.select(isWhiteBackgroundSelector).subscribe(this.setBackgroundColorToState)
    }

    private setBackgroundColorToState = (isWhiteBackground: boolean) => {
        ThreeRendererService.CLEAR_COLOR = isWhiteBackground
            ? ThreeRendererService.BACKGROUND_COLOR.white
            : ThreeRendererService.BACKGROUND_COLOR.normal
        this.renderer?.setClearColor(ThreeRendererService.CLEAR_COLOR, ThreeRendererService.CLEAR_ALPHA)
    }

    private initGL = (containerWidth: number, containerHeight: number) => {
        this.setGLOptions()
        this.renderer = new WebGLRenderer(ThreeRendererService.RENDER_OPTIONS)
        if (ThreeRendererService.setPixelRatio) {
            this.renderer.setPixelRatio(window.devicePixelRatio)
        }
        if (ThreeRendererService.enableFXAA) {
            if (WEBGL.isWebGL2Available) {
                const size = this.renderer.getDrawingBufferSize(new Vector2())
                const renderTarget = new WebGLRenderTarget(size.width, size.height, {
                    format: RGBAFormat
                })
                this.composer = new CustomComposer(this.renderer, renderTarget)
            } else {
                this.composer = new CustomComposer(this.renderer)
            }
        }
        this.renderer.setSize(containerWidth, containerHeight)
        this.renderer.domElement.id = "codeMapScene"

        if (ThreeRendererService.enableFXAA) {
            this.initComposer()
        }
    }

    private setGLOptions = () => {
        switch (this.state.getValue().appSettings.sharpnessMode) {
            case SharpnessMode.Standard:
                ThreeRendererService.RENDER_OPTIONS.antialias = true
                ThreeRendererService.enableFXAA = false
                ThreeRendererService.setPixelRatio = false
                break
            case SharpnessMode.PixelRatioNoAA:
                ThreeRendererService.RENDER_OPTIONS.antialias = false
                ThreeRendererService.enableFXAA = false
                ThreeRendererService.setPixelRatio = true
                break
            case SharpnessMode.PixelRatioFXAA:
                ThreeRendererService.RENDER_OPTIONS.antialias = false
                ThreeRendererService.enableFXAA = true
                ThreeRendererService.setPixelRatio = true
                break
            case SharpnessMode.PixelRatioAA:
                ThreeRendererService.RENDER_OPTIONS.antialias = true
                ThreeRendererService.enableFXAA = false
                ThreeRendererService.setPixelRatio = true
                break
        }
    }

    private initComposer = () => {
        const pixelRatio = this.renderer.getPixelRatio()

        this.composer.setSize(window.innerWidth * pixelRatio, window.innerHeight * pixelRatio)
        const renderPass = new RenderPass(this.scene, this.camera)
        this.composer.addPass(renderPass)

        const effectFXAA = new ShaderPass(new fxaaShaderStrings())
        effectFXAA.renderToScreen = false
        effectFXAA.uniforms["resolution"].value.x = 1 / (window.innerWidth * pixelRatio)
        effectFXAA.uniforms["resolution"].value.y = 1 / (window.innerHeight * pixelRatio)
        this.composer.addPass(effectFXAA)
    }

    getInfo = (): WebGLInfo["render"] => {
        return ThreeRendererService.enableFXAA ? this.composer.getInfo() : this.renderer.info.render
    }

    getMemoryInfo = (): WebGLInfo["memory"] => {
        return ThreeRendererService.enableFXAA ? this.composer.getMemoryInfo() : this.renderer.info.memory
    }

    render() {
        const { scene, camera, composer, renderer } = this
        if (ThreeRendererService.enableFXAA) {
            composer?.render()
        } else {
            renderer?.render(scene, camera)
        }
    }
}
