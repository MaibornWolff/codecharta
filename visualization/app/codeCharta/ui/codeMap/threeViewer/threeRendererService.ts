import { WebGLRenderer } from "three"
import { IRootScopeService } from "angular"
import { StoreService } from "../../../state/store.service"
import {
	IsWhiteBackgroundService,
	IsWhiteBackgroundSubscriber
} from "../../../state/store/appSettings/isWhiteBackground/isWhiteBackground.service"

export class ThreeRendererService implements IsWhiteBackgroundSubscriber {
	static BACKGROUND_COLOR = {
		white: 0xffffff,
		normal: 0xeeeedd
	}

	static CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.normal

	static CLEAR_ALPHA = 1

	static RENDER_OPTIONS = {
		antialias: true,
		preserveDrawingBuffer: true
	}

	renderer: WebGLRenderer

	constructor(private storeService: StoreService, private $rootScope: IRootScopeService) {}

	init(containerWidth: number, containerHeight: number) {
		this.renderer = new WebGLRenderer(ThreeRendererService.RENDER_OPTIONS)
		IsWhiteBackgroundService.subscribe(this.$rootScope, this)
		this.renderer.setSize(containerWidth, containerHeight)
		this.onIsWhiteBackgroundChanged(this.storeService.getState().appSettings.isWhiteBackground)
	}

	onIsWhiteBackgroundChanged(isWhiteBackground: boolean) {
		if (isWhiteBackground) {
			ThreeRendererService.CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.white
		} else {
			ThreeRendererService.CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.normal
		}
		this.renderer.setClearColor(ThreeRendererService.CLEAR_COLOR, ThreeRendererService.CLEAR_ALPHA)
	}
}
