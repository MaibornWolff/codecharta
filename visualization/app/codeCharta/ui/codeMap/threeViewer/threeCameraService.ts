"use strict"

import { PerspectiveCamera, Vector3 } from "three"
import { IRootScopeService } from "angular"
import throttle from "lodash.throttle"
import { CameraChangeSubscriber, ThreeOrbitControlsService } from "./threeOrbitControlsService"
import { StoreService } from "../../../state/store.service"
import { setCamera } from "../../../state/store/appSettings/camera/camera.actions"
import { CameraService, CameraSubscriber } from "../../../state/store/appSettings/camera/camera.service"

export class ThreeCameraService implements CameraChangeSubscriber, CameraSubscriber {
	static VIEW_ANGLE = 45
	static NEAR = 1 // !NOTE decreased in order to disable clipping for street layout, can be reverted after rescaling and repositioning the layout
	static FAR = 200000 //TODO optimize renderer for far objects
	private static DEBOUNCE_TIME = 400
	private readonly throttledCameraChange: () => void

	camera: PerspectiveCamera

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		this.throttledCameraChange = throttle(() => {
			this.storeService.dispatch(setCamera(this.camera.position), { silent: true })
		}, ThreeCameraService.DEBOUNCE_TIME)
		CameraService.subscribe(this.$rootScope, this)
	}

	onStoreCameraChanged(camera: Vector3) {
		this.camera.position.set(camera.x, camera.y, camera.z)
		this.camera.lookAt(0, 0, 0)
	}

	onCameraChanged() {
		this.throttledCameraChange()
	}

	init(containerWidth: number, containerHeight: number) {
		const aspect = containerWidth / containerHeight
		this.camera = new PerspectiveCamera(ThreeCameraService.VIEW_ANGLE, aspect, ThreeCameraService.NEAR, ThreeCameraService.FAR)
		this.setPosition()
		ThreeOrbitControlsService.subscribe(this.$rootScope, this)
	}

	setPosition() {
		const { camera } = this.storeService.getState().appSettings
		this.camera.position.set(camera.x, camera.y, camera.z)
	}
}
