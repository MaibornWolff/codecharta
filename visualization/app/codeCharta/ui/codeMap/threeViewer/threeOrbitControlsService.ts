"use strict";
import {ThreeCameraService} from "./threeCameraService";
import {IRootScopeService, IAngularEvent} from "angular";
import {Object3D, OrbitControls, PerspectiveCamera} from "three";
import * as THREE from "three";
import {ThreeSceneService} from "./threeSceneService";

export interface CameraChangeSubscriber {
    onCameraChanged(camera: PerspectiveCamera, event: IAngularEvent);
}

/**
 * Service to manage the three orbit controls in an angular way.
 */
class ThreeOrbitControlsService {

    public static SELECTOR = "threeOrbitControlsService";
    public static CAMERA_CHANGED_EVENT_NAME = "camera-changed";

    controls: OrbitControls;

    /* ngInject */
    constructor(
        private threeCameraService: ThreeCameraService,
        private threeSceneService: ThreeSceneService,
        private $rootScope: IRootScopeService
    ) {}

    autoFitTo( obj = this.threeSceneService.mapGeometry) {

        const boundingSphere = new THREE.Box3().setFromObject( obj ).getBoundingSphere();

        const scale = 1.4; // object size / display size
        const objectAngularSize = ( this.threeCameraService.camera.fov * Math.PI / 180 ) * scale;
        const distanceToCamera = boundingSphere.radius / Math.tan( objectAngularSize / 2 );
        const len = Math.sqrt( Math.pow( distanceToCamera, 2 ) + Math.pow( distanceToCamera, 2 ) );

        this.threeCameraService.camera.position.set(len, len, len);
        this.controls.update();

        let t = boundingSphere.center.clone();
        t.setY(0);
        this.threeCameraService.camera.lookAt( t );
        this.controls.target.set( t.x, t.y, t.z );

        this.threeCameraService.camera.updateProjectionMatrix();

    }

    subscribe(subscriber: CameraChangeSubscriber) {
        this.$rootScope.$on(ThreeOrbitControlsService.CAMERA_CHANGED_EVENT_NAME, (event: IAngularEvent, camera: PerspectiveCamera) => {
            subscriber.onCameraChanged(camera, event);
        });
    }

    /**
     * Inits the controls on the given DOM Element
     * @param domElement Element with the canvas on it
     */
    init(domElement){
        const OrbitControls = require("three-orbit-controls")(require("three"));
        this.controls = new OrbitControls(this.threeCameraService.camera, domElement);
        let ctx= this;
        this.controls.addEventListener( "change", function () {
            ctx.onInput(ctx.threeCameraService.camera);
        });
    }

    /**
     * Called when the orbit controls receive an user input
     * @param {Camera} camera
     */
    onInput(camera: PerspectiveCamera) {
        this.$rootScope.$broadcast(ThreeOrbitControlsService.CAMERA_CHANGED_EVENT_NAME, camera);
    }

}

export {ThreeOrbitControlsService};




