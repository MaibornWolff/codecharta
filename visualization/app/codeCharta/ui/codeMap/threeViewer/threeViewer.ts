import {ThreeViewerService} from "./threeViewerService";
import {ThreeSceneService} from "./threeSceneService";
import {ThreeCameraService} from "./threeCameraService";
import {ThreeOrbitControlsService} from "./threeOrbitControlsService";
import {ThreeRendererService} from "./threeRendererService";
import {ThreeUpdateCycleService} from "./threeUpdateCycleService";
import "../../../core/settings/settings.module";

import angular from "angular";

angular.module("app.codeCharta.ui.codeMap.threeViewer", ["app.codeCharta.core.settings"])
    .service(
        ThreeViewerService.SELECTOR,
        ThreeViewerService
    )
    .service(
        ThreeUpdateCycleService.SELECTOR,
        ThreeUpdateCycleService
    )
    .service(
        ThreeSceneService.SELECTOR,
        ThreeSceneService
    )
    .service(
        ThreeRendererService.SELECTOR,
        ThreeRendererService
    )
    .service(
        ThreeOrbitControlsService.SELECTOR,
        ThreeOrbitControlsService
    )
    .service(
        ThreeCameraService.SELECTOR,
        ThreeCameraService
    );


