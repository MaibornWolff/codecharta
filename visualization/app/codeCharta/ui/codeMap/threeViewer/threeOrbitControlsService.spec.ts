import "./threeViewer";
import {NG} from "../../../../../mocks/ng.mockhelper";

/**
 * @test {ThreeOrbitControlsService}
 */
describe("app.codeCharta.ui.codeMap.threeViewer.threeOrbitControlsService", function() {

    //noinspection TypeScriptUnresolvedVariable
    beforeEach(NG.mock.module("app.codeCharta.ui.codeMap.threeViewer"));

    //noinspection TypeScriptUnresolvedVariable
    it("should retrieve the angular service instance", NG.mock.inject(function(threeOrbitControlsService){
        expect(threeOrbitControlsService).not.toBe(undefined);
    }));

});