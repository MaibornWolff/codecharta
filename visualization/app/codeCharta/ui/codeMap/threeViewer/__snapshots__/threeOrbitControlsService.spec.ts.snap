// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeOrbitControlsService rotateCameraInVectorDirection  1`] = `
Vector3 {
  "x": 0,
  "y": -1.855914808404301,
  "z": 0.00018559148084032714,
}
`;
